### PR TITLE
Feature/add chart widgets

### PR DESCRIPTION
**Pull request by lunaticcoding**
Once you reviewed this code I'll start adding the other types of graph. :)

### CODE REVIEW

**Comment by lunaticcoding** [comment]
Screenshot 2020-04-23 at 03 50 24

File: pdf/lib/widgets/chart.dart [hunk 1]
```diff
@@ -0,0 +1,204 @@
...
+    );
+
+    var gridLeft = box.left + maxTextWidth + xAxisMargin;
+    var gridTop = box.top - textHeight / 2;
+    var gridRight = box.right - maxTextWidth / 2;
+    var gridBottom = box.bottom + textHeight + yAxisMargin;
+
+    if (pointLine) {
+      var lastPoint = 0.0;
+      data.asMap().forEach((int i, double point) {
```
Comment (DavBfr) [inline pdf/lib/widgets/chart.dart:177]: To optimize the drawing operations and reduce the size of the Pdf, you can move the line settings out of the loop:

```dart
      data.asMap().forEach((int i, double point) {
        context.canvas.drawLine(
            gridLeft + (gridRight - gridLeft) * i / (xAxis.length - 1),
            gridBottom + (gridTop - gridBottom) * lastPoint / maxValue,
            gridLeft + (gridRight - gridLeft) * (i + 1) / (xAxis.length - 1),
            gridBottom + (gridTop - gridBottom) * point / maxValue);
        lastPoint = point;
      });
      context.canvas
        ..setStrokeColor(pointLineColor)
        ..setLineWidth(pointLineWidth)
        ..setLineCap(PdfLineCap.joinRound)
        ..setLineJoin(PdfLineCap.joinRound)
        ..strokePath();
```

I also add `setLineCap` and `setLineJoin`, it renders nicer lines.

File: pdf/lib/widgets/chart.dart [hunk 2]
```diff
@@ -0,0 +1,204 @@
...
+      textHeight,
+    );
+
+    var gridLeft = box.left + maxTextWidth + xAxisMargin;
+    var gridTop = box.top - textHeight / 2;
+    var gridRight = box.right - maxTextWidth / 2;
+    var gridBottom = box.bottom + textHeight + yAxisMargin;
+
+    if (pointLine) {
+      var lastPoint = 0.0;
```
Comment (DavBfr) [inline pdf/lib/widgets/chart.dart:176]: this first point is not true. Should be null and in the loop, do nothing if lastPoint is not set.

File: pdf/lib/widgets/chart.dart (hunk 1 above)
Comment (lunaticcoding) [reply pdf/lib/widgets/chart.dart:177]: Ah you're right :O Good point thanks :) I'll fix it tonight.

File: pdf/lib/widgets/chart.dart (hunk 2 above)
Comment (lunaticcoding) [reply pdf/lib/widgets/chart.dart:176]: To clarify: You mean simply use lastPoint as both the starting point and the boolean here? 
so 

if(lastPoint != null {
    loop
}

Comment (DavBfr) [reply pdf/lib/widgets/chart.dart:176]: No:
loop {
  if(lastPoint != null {
     draw line
  }
   lastpoint=point
}

File: pdf/lib/widgets/chart.dart (hunk 1 above)
Comment (lunaticcoding) [reply pdf/lib/widgets/chart.dart:177]: Does the same apply here? 

      data.asMap().forEach((int i, double point) {
        context.canvas
          ..setColor(pointColor)
          ..drawEllipse(
              gridLeft + (gridRight - gridLeft) * (i + 1) / (xAxis.length - 1),
              gridBottom + (gridTop - gridBottom) * point / maxValue,
              pointSize,
              pointSize)
          ..fillPath();
      });

as the following 

      data.asMap().forEach((int i, double point) {
        context.canvas.drawEllipse(
              gridLeft + (gridRight - gridLeft) * (i + 1) / (xAxis.length - 1),
              gridBottom + (gridTop - gridBottom) * point / maxValue,
              pointSize,
              pointSize)
      });

        context.canvas
          ..setColor(pointColor)
          ..fillPath();

File: pdf/lib/widgets/chart.dart (hunk 2 above)
Comment (lunaticcoding) [reply pdf/lib/widgets/chart.dart:176]: So I changed a few things. It's now possible to decide whether you want to draw only lines or only dots or both. (With the first dot being optional like you did in the code snippet you just posted.)

**Comment by lunaticcoding** [comment]
The "A better API would take an Axis object and a list of DataSet objects." is a very good point. :) Do you have some specific idea or is it okay if I make a proposal :D?

Since the company I currently work for would urgently need this feature would it be okay to make this API redesign a separate PR? Also if you have some resources for me to learn (about API design) I would love to hear about them.
Let me know :)

File: pdf/lib/widgets/chart.dart [hunk 3]
```diff
@@ -0,0 +1,216 @@
...
+        ..setLineWidth(pointLineWidth)
+        ..setLineCap(PdfLineCap.joinRound)
+        ..setLineJoin(PdfLineCap.joinRound)
+        ..strokePath();
+    }
+
+    if (drawPoints) {
+      data.asMap().forEach((int i, double point) {
+        context.canvas
+          ..setColor(pointColor)
```
Comment (DavBfr) [inline pdf/lib/widgets/chart.dart:204]: Same thing here: setColor and fillPath could be after the loop.

**Comment by DavBfr** [comment]
For a basic API redesign, I would have:

```dart
  Chart(
    width: 500,
    height: 250,
    axes: LinearAxes(
      xAxis: <double>[0, 1, 2, 3, 4, 5, 6],
      yAxis: <double>[0, 3, 6, 9],
    ),
    data: <DataSet>[
      ScatterDataSet(
        data: <double>[1, 3, 7],
        pointColor: PdfColors.green,
        pointSize: 8,
        lineColor: PdfColors.blue,
        lineWidth: 6,
      ),
    ],
  );
```

This way, you can have only one generic `Chart` object as the entry point.
It's possible to use another `Axes` descendent that implements logarithmic or dates, etc.
You can draw multiple `DataSet` on the same chart, with different colors or style: one Bars, one Line and one Scatter.

... learn about API design, I'm afraid that's mostly experience and looking at other's code, exchanging ideas.

**Comment by lunaticcoding** [comment]
Oh that looks great :) I'll get right to it :)

File: pdf/lib/widgets/chart.dart (hunk 3 above)
Comment (lunaticcoding) [reply pdf/lib/widgets/chart.dart:204]: yep got it :D

**Comment by lunaticcoding** [comment]
@DavBfr The assignment of PdfRect.fromLTRB() flips top and bottom :S am allowed to change this? Or should I create a second method and rename this one? 

PdfRect.fromLTRB(
    ...
    top: box.bottom + someOffsetValue
    ...
) 

Screenshot 2020-04-24 at 16 10 58
Screenshot 2020-04-24 at 16 11 05

**Comment by DavBfr** [comment]
I know that's strange, but in PDF coordinates "bottom" goes up

**Comment by lunaticcoding** [comment]
@DavBfr oh alright :S got it :)

Btw I implemented your API design :D Looks great

**Comment by lunaticcoding** [comment]
I also renamed the "Scatter Chart" because it's actually a Line Chart ;D my bad

File: pdf/lib/widgets/chart.dart [hunk 4]
```diff
@@ -0,0 +1,227 @@
...
+    if (drawLine) {
+      double lastPoint = lineStartingPoint;
+      data.asMap().forEach((int i, double point) {
+        if (lastPoint != null) {
+          context.canvas.drawLine(
+              box.left + box.width * i / data.length,
+              box.bottom + box.height * lastPoint / maxValue,
+              box.left + box.width * (i + 1) / data.length,
+              box.bottom + box.height * point / maxValue);
+          lastPoint = point;
```
Comment (DavBfr) [inline pdf/lib/widgets/chart.dart:205]: Small bug here: lastPoint should be after '}'

File: pdf/test/widget_chart_test.dart [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+    test('Default LineChart', () {
+      pdf.addPage(Page(
+        build: (Context context) => Chart(
+          grid: LinearGrid(
+            xAxis: <double>[0, 1, 2, 3, 4, 5, 6],
+            yAxis: <double>[0, 3, 6, 9],
+          ),
+          data: <DataSet>[
+            LineDataSet(
+              data: <double>[1, 3, 7],
```
Comment (DavBfr) [inline pdf/test/widget_chart_test.dart:43]: I would expect the data to be a list of tuple, like:
```dart
data: <ChartValue>[ChartValue(0, 1), ChartValue(1, 3), ChartValue(2, 7)],
```
That will give you the full coordinates of each point in the cartesian coordinate system represented by the Grid object.

File: pdf/lib/widgets/chart.dart [hunk 6]
```diff
@@ -0,0 +1,227 @@
...
+  TextStyle style;
+  PdfFont font;
+  PdfFontMetrics xAxisFontMetric;
+  PdfFontMetrics yAxisFontMetric;
+  double separatorLineWidth;
+  PdfColor separatorColor;
+
+  @override
+  void layout(Context context, PdfRect box) {
+    style = textStyle ?? Theme.of(context).defaultTextStyle;
```
Comment (DavBfr) [inline pdf/lib/widgets/chart.dart:96]: style = Theme.of(context).defaultTextStyle.merge(textStyle);

File: pdf/lib/widgets/chart.dart [hunk 7]
```diff
@@ -0,0 +1,227 @@
...
+        box.bottom + xAxisFontMetric.height + yMargin,
+        box.right - xAxisFontMetric.width / 2,
+        box.top - yAxisFontMetric.height / 2);
+  }
+
+  @override
+  void paint(Context context, PdfRect box) {
+    xAxis.asMap().forEach((int i, double x) {
+      context.canvas
+        ..drawString(
```
Comment (DavBfr) [inline pdf/lib/widgets/chart.dart:117]: need to set the color here (style.color)

File: pdf/lib/widgets/chart.dart (hunk 4 above)
Comment (lunaticcoding) [reply pdf/lib/widgets/chart.dart]: already fixed ;D

File: pdf/test/widget_chart_test.dart (hunk 5 above)
Comment (lunaticcoding) [reply pdf/test/widget_chart_test.dart]: Done. (I also accounted for the grid starting at a negative value)

**Comment by lunaticcoding** [comment]
I also changed the names in ChartValue from xVal, yVal to simply x, y (I don't know if this shows up in your review because I requested the review before changing those names)

**Comment by DavBfr** [comment]
I will merge it like that.

I added gradient fill and dashed lines to the pdf library in the master branch. You will be able to do nice shaded bars and dashed grids!

```dart
context.canvas
      ..saveContext()
// draw any path, but don't fill it
      ..clipPath()
      ..applyShader(
        PdfShading(
          context.document,
          shadingType: PdfShadingType.axial,
          function: PdfFunction(
        context.document,
        colors: [PdfColors.red, PdfColors.blue],
      ),
          start: PdfPoint(10, 10),
          end: PdfPoint(200, 200),
          extendStart: true,
          extendEnd: true,
        ),
      )
      ..restoreContext();
```

**Pull request closed without merging**